src/results.rs

```diff
@@ -11,6 +12,23 @@ use std::str::FromStr;
 use toolchain::Toolchain;
 use util;
 
+pub trait ExperimentResultDB {
+    type CrateWriter: CrateResultWriter;
+    fn for_crate(&self, crate_: &ExCrate, toolchain: &Toolchain) -> Self::CrateWriter;
+
+    fn delete_all_results(&self) -> Result<()>;
+}
+
+pub trait CrateResultWriter {
+    /// Return a path fragement that can be used to identify this crate and
+    /// toolchain.
+    fn result_path_fragement(&self) -> PathBuf;
+
+    fn record_results<F>(&self, f: F) -> Result<TestResult> where F: FnOnce() -> Result<TestResult>;
+    fn get_test_results(&self) -> Result<Option<TestResult>>;
```

Comment:
should this be plural? looks like it's only returning one result? though maybe `TestResult` is actually multiple results

also, getters in Rust tend to not start with `get_`, but not a big deal